src/main/java/net/fabricmc/loom/util/Version.java

```diff
@@ -0,0 +1,116 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Ordering;
+import org.gradle.util.internal.VersionNumber;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A simple version class that can be used to compare versions.
+ * This class allows for versions that are not strictly following the semver specification,
+ * but are still allowed in the context of gradle versioning.
+ *
+ * <p>This class is intentionally very flexible and does not enforce any specific versioning scheme,
+ * and should be very similar to the versioning used by gradle itself.
+ */
+public record Version(int major, int minor, int micro, int patch, @Nullable String qualifier) implements Comparable<Version> {
+	private static final Pattern REGEX = Pattern.compile("(\\d+)(?:\\.(\\d+))?(?:\\.(\\d+))?()(?:[.-]([^+\\s]*))?(?:\\+.*)?");
+	private static final Pattern REGEX_WITH_PATCH = Pattern.compile("(\\d+)(?:\\.(\\d+))?(?:\\.(\\d+))?(?:\\.(\\d+))?(?:[+.-]([^+\\s]*))?(?:\\+.*)?");
+	public static final Version UNKNOWN = new Version(0, 0, 0, 0, null);
+
+	public static Version parse(String version) {
+		return parse(version, false);
+	}
+
+	public static Version parse(String version, boolean withPatch) {
+		Matcher matcher = (withPatch ? REGEX_WITH_PATCH : REGEX).matcher(version);
+
+		if (!matcher.matches()) {
+			return UNKNOWN;
+		}
+
+		int major = Integer.parseInt(matcher.group(1));
+		int minor = !Strings.isNullOrEmpty(matcher.group(2)) ? Integer.parseInt(matcher.group(2)) : 0;
+		int micro = !Strings.isNullOrEmpty(matcher.group(3)) ? Integer.parseInt(matcher.group(3)) : 0;
+		int patch = !Strings.isNullOrEmpty(matcher.group(4)) ? Integer.parseInt(matcher.group(4)) : 0;
+		String qualifier = matcher.group(5);
+
+		return new Version(major, minor, micro, patch, qualifier);
+	}
+
+	public VersionNumber asBaseVersion() {
+		return new VersionNumber(this.major, this.minor, this.micro, this.patch, null);
+	}
+
+	@Override
+	public int compareTo(@NotNull Version other) {
+		if (this.major != other.major) {
+			return this.major - other.major;
+		} else if (this.minor != other.minor) {
+			return this.minor - other.minor;
+		} else if (this.micro != other.micro) {
+			return this.micro - other.micro;
+		} else {
+			return this.patch != other.patch ? this.patch - other.patch
+					: Ordering.natural().nullsLast()
+					.compare(this.toLowerCase(this.qualifier), this.toLowerCase(other.qualifier));
+		}
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj) {
+			return true;
+		} else if (obj == null || this.getClass() != obj.getClass()) {
+			return false;
+		} else {
+			Version other = (Version) obj;
+			return this.major == other.major && this.minor == other.minor && this.micro == other.micro && this.patch == other.patch
+					&& Objects.equals(this.toLowerCase(this.qualifier), this.toLowerCase(other.qualifier));
+		}
+	}
```

Comment:
I wonder if it'd just be better to convert the qualifier/build metadata to lowercase when `Version` is created and use the default equals/hashCode to make the code simpler. This is fine too though